### PR TITLE
Adds ecr login and refines code

### PR DESCRIPTION
**Pull request by scriptnull**
https://github.com/Shippable/execTemplates/issues/12

This PR does the following
- cleans up shell scripts in integrations
- adds ability to login and push/pull docker images to amazon ecr in a runSh job

# YMLs

#### shippable.resources.yml
```yml
resources:
  - name: awsKeys-for-ec2
    type: cliConfig
    integration: test-aws
    pointer:
      region: us-west-2

  - name: awsKeys-for-ecr
    type: cliConfig
    integration: test-aws
    pointer:
      region: us-west-2
```

#### shippable.jobs.yml
```yml
jobs:
  - name: amazonKeys-basic
    type: runSh
    steps:
      - IN: awsKeys-for-ec2
      - TASK:
        - script: echo "Listing running instances..."
        - script: aws ec2 describe-instances
        - script: echo "access key value -- $AWS_ACCESS_KEY -- should not be visible here"
        
  - name: amazonKeys-ecr
    type: runSh
    steps:
      - IN: awsKeys-for-ecr
        scopes:
          - ecr
      - TASK:
        - script: docker pull busybox:latest
        - script: docker tag busybox:latest 854172162208.dkr.ecr.us-west-2.amazonaws.com/checking:$BUILD_NUMBER
        - script: docker push 854172162208.dkr.ecr.us-west-2.amazonaws.com/checking:$BUILD_NUMBER
```

# Success

#### Reading list of EC2 machines
x16

#### Pushing to ECR repo
x15

# Failure

#### AWS credentials is wrong
x18

#### pointer.region is missing in resource
x17

### CODE REVIEW

File: integrations/amazonKeys/cli/init.sh [hunk 1]
```diff
@@ -1,51 +1,32 @@
...
-  AWS_ACCESS_KEY="$( shipctl get_integration_resource_field $RESOURCE_NAME "accessKey" )"
-  AWS_SECRET_KEY="$( shipctl get_integration_resource_field $RESOURCE_NAME "secretKey" )"
-  RESOURCE_VERSION_PATH="$(shipctl get_resource_meta $RESOURCE_NAME)/version.json"
-  AWS_REGION="$( shipctl get_json_value $RESOURCE_VERSION_PATH "propertyBag.yml.pointer.region" )"
+  AWS_ACCESS_KEY="$( shipctl get_integration_resource_field "$RESOURCE_NAME" "accessKey" )"
+  AWS_SECRET_KEY="$( shipctl get_integration_resource_field "$RESOURCE_NAME" "secretKey" )"
+  RESOURCE_VERSION_PATH="$(shipctl get_resource_meta "$RESOURCE_NAME")/version.json"
+  AWS_REGION="$( shipctl get_json_value "$RESOURCE_VERSION_PATH" "propertyBag.yml.pointer.region" )"
 
   if is_empty "$AWS_ACCESS_KEY"; then
```
Comment (ric03uec) [inline integrations/amazonKeys/cli/init.sh]: square brackets here.

File: README.md [hunk 2]
```diff
@@ -5,3 +5,17 @@ This repository will (eventually) contain all the scripts that are pieced togeth
 ## Integrations
 
 Under the integrations folder, you'll find the scripts that get executed when a particular integration gets used in a build.
+
+#### Contribute
+- create following file structure inside `integrations` directory
```
Comment (shrivara) [inline README.md]: Create the following...

File: README.md [hunk 3]
```diff
@@ -5,3 +5,17 @@ This repository will (eventually) contain all the scripts that are pieced togeth
...
+```
```
Comment (shrivara) [inline README.md]: Indent this by 4 spaces so its below the bullet points.

File: integrations/amazonKeys/cli/init.sh [hunk 4]
```diff
@@ -1,51 +1,32 @@
 #!/bin/bash -e
-readonly IFS=$'\n\t'
-readonly ROOT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
-readonly SCRIPT_NAME="$( basename "$0" )"
-readonly ARGS=("$@")
 
-export RESOURCE_NAME=""
-export SCOPES=""
+# shellcheck source=integrations/common/envs.sh
+source "$(dirname "$0")/../../common/envs.sh"
```
Comment (ric03uec) [inline integrations/amazonKeys/cli/init.sh]: define them with relative path

```
readonly ROOT_DIR=....
readonly COMMON_DIR=$ROOT_DIR/../../common
readonly ENVS="$COMMON_DIR/envs.sh
```

File: integrations/amazonKeys/cli/init.sh [hunk 5]
```diff
@@ -1,51 +1,32 @@
...
-        SCOPES=$2
-        ;;
-    esac
-  else
-    print_help
-    exit 1
-  fi
-}
+check_params() {
+  RESOURCE_NAME=${ARGS[0]}
```
Comment (ric03uec) [inline integrations/amazonKeys/cli/init.sh]: add `local` here

File: README.md [hunk 6]
```diff
@@ -5,3 +5,17 @@ This repository will (eventually) contain all the scripts that are pieced togeth
...
+integrationName/
+`-- cli
+    |-- cleanup.sh
+    `-- init.sh
+```
+- Write `init.sh` and `cleanup.sh`
```
Comment (shrivara) [inline README.md]: Lets expand on this:
- Populate `init.sh` with the script required to configure the integration.
- Populate `cleanup.sh` with the script required to clean up the integration.
- Use existing integration handler as references.

File: integrations/amazonKeys/cli/init.sh [hunk 7]
```diff
@@ -1,51 +1,32 @@
...
 
-export RESOURCE_NAME=""
-export SCOPES=""
+# shellcheck source=integrations/common/envs.sh
+source "$(dirname "$0")/../../common/envs.sh"
+# shellcheck source=integrations/common/utils.sh
+source "$(dirname "$0")/../../common/utils.sh"
 
-print_help() {
+RESOURCE_NAME=""
```
Comment (ric03uec) [inline integrations/amazonKeys/cli/init.sh]: add export here

File: README.md [hunk 8]
```diff
@@ -5,3 +5,17 @@ This repository will (eventually) contain all the scripts that are pieced togeth
...
+- Lint your scripts using [shellcheck](https://github.com/koalaman/shellcheck)
+```bash
```
Comment (shrivara) [inline README.md]: Indent this by 4 spaces so its below the bullet points.

File: integrations/amazonKeys/cli/init.sh [hunk 9]
```diff
@@ -63,37 +44,45 @@ check_and_set_vars () {
...
-    echo "Found scopes: $SCOPES"
-  fi
-
-  _configure_aws_cli
-
-  if has_scope "ecr"; then
-    _configure_aws_ecr
+  check_params
+  init_scope_configure
+  if csv_has_value "$SCOPES" "ecr"; then
```
Comment (ric03uec) [inline integrations/amazonKeys/cli/init.sh]: square bracket

File: integrations/amazonKeys/cli/init.sh [hunk 10]
```diff
@@ -63,37 +44,45 @@ check_and_set_vars () {
...
+    init_scope_ecr
```
Comment (ric03uec) [inline integrations/amazonKeys/cli/init.sh]: add a convention here
- any function NOT called from main should start with underscore

**Review by shrivara — DISMISSED** [review]

File: integrations/common/envs.sh [hunk 11]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash -e
+
+readonly IFS=$'\n\t'
```
Comment (ric03uec) [inline integrations/common/envs.sh]: not sure why this file is required, cant we just push these chances into `utils.sh`.
- also utils.sh should be called _helpers.sh to make this consistent with admiral syntax

File: integrations/common/utils.sh [hunk 12]
```diff
@@ -0,0 +1,28 @@
+#!/bin/bash -e
+
+# is_empty checks if given value is "" or "null"
+# args: string
+is_empty() {
```
Comment (ric03uec) [inline integrations/common/utils.sh]: all functions here need tos tart with underscore

Comment (ric03uec) [reply integrations/common/utils.sh]: add functions to log success, error and markers.

**Review by ric03uec — COMMENTED** [review]
- overall, there's a lot that needs to be done here. we need to follow existing standards and not add code that is doing completely new things
- second biggest problem here is that there is no verbosity in the scripts. There's no way to print decent log messages and the existing scripts are not even doing a basic echo "something is happening". its like a black box that'll fail without any context and at THAT time we'll be trying to make guesses as to where something is going wrong.

**Comment by shrivara** [comment]
@ric03uec, here are a few things we discussed yesterday with regard to the scripts:
- Verbosity: We wanted to keep the output as clean as possible when there are no errors, so the console logs is not cluttered by it. Only when there are errors will they show up.
- Underscores in function names: We discussed that this is a source of confusion as to when something should have an underscore and what level of nesting is present. Maybe adding it it for helper methods has some value, but I'm not sure about the main scripts itself.

For the rest of the comments:
> add export here

We wanted to keep the scope of the variable as little as possible.

> square bracket

It seemed cleaner to do this without square brackets. I guess it could be argued that we need square brackets to keep checks consistent though.

Thoughts?

**Comment by ric03uec** [comment]
- verbosity is needed here just like in build scripts. the reason we show everything thats going on in build scripts is because it makes thats the reason for showing console logs to the user in first place. Its their job and we should tell them whats going on. Especially in bash which is where the execution is happening.  The thought here should be "show everything unless something needs to be hidden" rather than "hide everything unless something needs to be shown". The least i want to know is what steps are happening. this should not be a black box that either runs without any messages or throws an error. whats the point of showing the section itself if we dont want to show whats going on inside it
- nott sure how export will cause any issues. when the variable is put outside the function, its global as it is. i get a cleaner syntax with export
- lets keep square brackets consistent. I want us to treat this as our internal codebase and everything should follow a standard. if you want to remove square brackets here, then they have to go out from the rest of the system
-

**Comment by ric03uec** [comment]
for underscore vs not underscore, lets go with your recommendation to make it less confusing. all helpers start with underscore and in-file calls are without it

**Comment by scriptnull** [comment]
@ric03uec I have made most of the changes and updated the screenshots. But I would like to show you the look and feel of "not having brackets vs having brackets" in if conditions in some places.

## Implementation 1
__Definition__:
```bash
_is_empty() {
  [[ -z "$1" ]] || [[ "$1" == "null" ]]
}
```

__Usage__:
```bash
export EMPTY_STRING=""
if _is_empty "$EMPTY_STRING"; then
  echo "EMPTY_STRING variable is empty"
fi
```

## Implementation 2
__Definition__:
```bash
_is_empty() {
  [[ -z "$1" ]] || [[ "$1" == "null" ]]
  echo $?
}
```

__Usage__:
```bash
export EMPTY_STRING=""
if [[ "$(_is_empty "$EMPTY_STRING")" == 0 ]]; then
  echo "EMPTY_STRING variable is empty"
fi
```

## Analysis
Implementation 2 seems bit unclear when compared to implementation 1 due to some reasons
1. Implementation 1 behaves more shell like, "by defining shell functions like commands (output is conveyed via exit status code )"
1. Readability between `_is_empty "$EMPTY_STRING";` vs `[[ "$(_is_empty "$EMPTY_STRING")" == 0 ]];`


Please suggest if there is a more readable way of a implementation which can have `[[` `]]` but not the cryptic check like `== 0` in it.

Remaining code review points are already addressed. So, if you are ok with not having `[[` `]]` in if conditions in some places (due to above analysis), this PR is up for merge.

**Comment by ric03uec** [comment]
- the problem hereis that the interface of is_empty is too simple and most of the code is still duplicated. im not really getting anything out of adding that function. wouldn't adding something like this make more sense
```
_check_env() {
  local env_key="$1"
  if [ -z "$env_key" ]; then
    __process_error "ERROR !!! Missing ENV variable: $env_key"
    exit 1
  elif [ "$env_key" == "" ]; then
    __process_error "ERROR !!! Empty value in ENV variable: $env_key"
    exit 1
  else
    __process_success "Found ENV: $env_key"
  fi
}

```
- this actually abstracts out a piece of logic we're going to use in every single file and give a much robust checking of envs

**Comment by ric03uec** [comment]
@shrivara @scriptnull the main point is that since we're goin gto use this function everwhere, we can ehnance it to have more functionality than simply doing an `if `condition check.

File: integrations/amazonKeys/cli/init.sh [hunk 13]
```diff
@@ -1,99 +1,106 @@
 #!/bin/bash -e
-readonly IFS=$'\n\t'
```
Comment (ric03uec) [inline integrations/amazonKeys/cli/init.sh:2]: - why is this removed? .we need it for more predictable iteration of the arrrays

Comment (scriptnull) [reply integrations/amazonKeys/cli/init.sh:2]: We now have this in `common/_helpers.sh` ( https://github.com/Shippable/execTemplates/pull/16/files/d3d2c8c3206fa381c90c07050f78da0283453498#diff-569876d2c3cb06a576a7d4506fd4537fR3 ), since it is going to be sourced everywhere and we wanted this variable everywhere.

**Comment by shrivara** [comment]
@ric03uec, while I think we should a `_check_env` method as you suggested, we should probably also keep `_is_empty`. This is quite similar to what we're used to with the underscore library which provides `.isEmpty`, this is simply easier to understand than `-z` checks everywhere and we'll also ensure that null checks are not missed. I think `_check_env` can use `_is_empty` internally.

**Comment by shrivara** [comment]
@ric03uec, merging this as the functionality is in place. Please let us know the styling changes that you would like and we will incorporate this later.
